src/Utilities/RandomGenerator.h

```diff
@@ -30,118 +30,46 @@
-      a[i]     = sqrt(-2.0 * log(temp1)) * cos(6.283185306 * temp2);
-      a[i + 1] = sqrt(-2.0 * log(temp1)) * sin(6.283185306 * temp2);
-    }
-    if (n % 2 == 1)
-    {
-      double temp1 = 1 - 0.9999999999 * rng(), temp2 = rng();
-      a[n - 1] = sqrt(-2.0 * log(temp1)) * cos(6.283185306 * temp2);
-    }
-  }
-
-  template<typename RNG>
-  static inline void generate(RNG& rng, float* restrict a, int n)
-  {
-    for (int i = 0; i + 1 < n; i += 2)
-    {
-      float temp1 = 1.0f - 0.9999999999f * rng(), temp2 = rng();
-      a[i]     = sqrtf(-2.0f * logf(temp1)) * cosf(6.283185306f * temp2);
-      a[i + 1] = sqrtf(-2.0f * logf(temp1)) * sinf(6.283185306f * temp2);
-    }
-    if (n % 2 == 1)
-    {
-      float temp1 = 1.0f - 0.9999999999f * rng(), temp2 = rng();
-      a[n - 1] = sqrtf(-2.0f * logf(temp1)) * cosf(6.283185306f * temp2);
-    }
-  }
-};
-
-inline uint32_t make_seed(int i, int n) { return static_cast<uint32_t>(std::time(0)) % 10474949 + (i + 1) * n + i; }
-
+#include <cstdint>
 // The definition of the fake RNG should always be available for unit testing
 #include "Utilities/FakeRandom.h"
-#ifdef USE_FAKE_RNG
-namespace qmcplusplus
-{
-typedef FakeRandom RandomGenerator_t;
-
-} // namespace qmcplusplus
-#else
-
 #ifdef HAVE_LIBBOOST
-
 #include "Utilities/BoostRandom.h"
-namespace qmcplusplus
-{
-template<class T>
-using RandomGenerator   = BoostRandom<T>;
-using RandomGenerator_t = BoostRandom<OHMMS_PRECISION_FULL>;
-} // namespace qmcplusplus
 #else
-
 #error -DHAVE_LIBBOOST is missing in the compile line. A cmake dependency fix is needed.
-
-#endif
 #endif
 
+uint32_t make_seed(int i, int n);
 
 namespace qmcplusplus
 {
-class RNGThreadSafe : public RandomGenerator_t
+template<class RNG>
+class RNGThreadSafe : public RNG
 {
 public:
-  inline RandomGenerator_t::result_type rand()
-  {
-    result_type result;
-// This should be a named section but at least clang 9 doesn't seem to support
-// and warns of extra tokens.
-#pragma omp critical
-    {
-      result = RandomGenerator_t::rand();
-    }
-    return result;
-  }
+  using result_type = typename RNG::result_type;
+
+  result_type rand();
 
   /** return a random number [0,1)
    */
-  inline RandomGenerator_t::result_type operator()()
-  {
-    result_type result;
-#pragma omp critical
-    {
-      result = RandomGenerator_t::rand();
-    }
-    return result;
-  }
+  result_type operator()();
 
-  /** generate a series of random numbers */
-  template<typename T1>
-  inline void generate_uniform(T1* restrict d, int n)
-  {
-#pragma omp critical
-    {
-      for (int i = 0; i < n; ++i)
-        d[i] = RandomGenerator_t::rand();
-    }
-  }
 };
 
-extern RNGThreadSafe Random;
+extern template class RNGThreadSafe<FakeRandom>;
+extern template class RNGThreadSafe<BoostRandom<float>>;
+extern template class RNGThreadSafe<BoostRandom<double>>;
 
-} // namespace qmcplusplus
+#ifdef USE_FAKE_RNG
+using RandomGenerator_t = FakeRandom;
```

Comment:
I'd really like to see this behavior by typedef dropped.  But that is Outside of this scope.

Reply:
Which line has typedef?

Reply:
65, Well type alias, but basically it uses the c preprocessor + linking to a different definition of the type to fake a template class. In much new code I've just made the RandomGenerator a function argument and its type a function template parameter (I honestly prefer this). Then no USE_FAKE_RNG defines are needed.  The RNG shouldn't be state for anything other than the crowd/thread context.

I don't think anything here makes that harder, and we have many existing tests that depend on this type alias swap.